Explanation of the role of the customer and his id (loyalty card)

elena-yevs opened the pull request:
Adding class Customer with constructor that takes LoyaltyCard as a parameter.

elena-yevs committed:
Initial commit

elena-yevs committed:
creating class Customer

elena-yevs committed:
creating class Customer in a new branch;
adding a constructor

elena-yevs committed:
class object creation

elena-yevs commented:
@alexander-yevsyukov unfortunately, “Reviewers” section doesn`t want to offer me a user after I enter a nickname, so for now I am addressing you like this.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,9 @@
+package com.example;
+
+public class Customer {
```
Do you intend to have classes derived from `Customer`? If not, the class needs to be made `final`.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,9 @@
+package com.example;
+
+public class Customer {
+    private String LoyaltyCard;
```
Please review the standard style conventions for Java. Field names start from lowercase letter.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,9 @@
+public class Customer {
+    private String LoyaltyCard;
+
+    public Customer(String LoyaltyCard) {
```
The same rule for lowercase names is used for parameter names.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,9 @@
+    private String LoyaltyCard;
+
+    public Customer(String LoyaltyCard) {
+        this.LoyaltyCard = LoyaltyCard;
```
What you need to have is another class called `LoyaltyCard`. Then, objects of this class would be initialized with a string containing a card number.

alexander-yevsyukov reviewed (changes requested):
Please see my comments and questions.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,9 @@
+package com.example;
+
```
Here should be Javadoc. Every `public`, `protected`, and package-private thing in a code _must_ be documented. When you show your code to colleagues it _needs_ documentation. It's written and maintained together with the executable code. It _is_ an integral part of the code.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,9 @@
+
+public class Customer {
+    private String LoyaltyCard;
+
```
Here should be Javadoc too.

alexander-yevsyukov replied:
I mean it.

alexander-yevsyukov commented on src/main/java/com/example/Main.java:
```diff
@@ -2,6 +2,6 @@
 
 public class Main {
```
Here should be Javadoc. What does this class do? What is it for?

elena-yevs committed:
parameter is written in lowerCamelCase.

elena-yevs committed:
class Customer is final

elena-yevs committed:
adding Javadoc for Main class

elena-yevs committed:
adding @author and @version in Javadoc for Main class

elena-yevs committed:
Javadoc for Customer class;
adding information in Javadoc of Main class

elena-yevs committed:
formatting Javadocs

elena-yevs committed:
Changing Javadoc for Customer.java:
1. new description for class Customer.
2. loyaltyCard is final.
3. editing Javadoc for constructor.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,23 @@
+
+/**
+ * Public class Customer with the field loyaltyCard and
+ * constructor
```
This documentation does not add any value. I can see that from the code. Do Not Repeat Yourself (DRY).

You need to describe the class from a modeling perspective. Something like: “A customer is a person who visits a shop for buying goods.” Then, add something about possession of a loyalty card.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,23 @@
+ * Public class Customer with the field loyaltyCard and
+ * constructor
+ *
+ * @author Elena Yevsyukova
```
Even though there is such a tag, software is a collective enterprise. Tagging code with authorship prevents collective ownership and hinders adding modifications. So, in modern days this tag is deprecated. Please remove.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,23 @@
+ * constructor
+ *
+ * @author Elena Yevsyukova
+ * @version 2.1
```
No, it's not version 2.1. You have not released this code to public ever. So it's pre 1.0. Forget about this tag for the time being.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,23 @@
+ */
+public final class Customer {
+    /**
+     *The private field of type String
```
Of course, it is. I can see it from the code! Why are we having this field? What does it mean from the domain perspective?

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,23 @@
+    /**
+     *The private field of type String
+     */
+    private String loyaltyCard;
```
This field needs to be `final` and of different type.

elena-yevs replied:
Why does the field need to be of a different type?

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,23 @@
+
+    /**
+     * Constructor takes loyaltyCard as a parameter
+     * @param loyaltyCard String value
```
I can see that from the code. Write something meaningful.

alexander-yevsyukov commented on src/main/java/com/example/Main.java:
```diff
@@ -1,7 +1,21 @@
 package com.example;
 
+/**
+ * Main class that contains void main
```
Finish sentences with periods. Again, I can see what you write in the code. What does this class do? Why are we having it?

alexander-yevsyukov commented on src/main/java/com/example/Main.java:
```diff
@@ -1,7 +1,21 @@
+    /**
+     * Here starts point of the program
+     *
+     * @param args command line values
```
I know this from the convention about the `main()` method. What parameters does it accept? How do I use them?

alexander-yevsyukov commented on src/main/java/com/example/Main.java:
```diff
@@ -1,7 +1,21 @@
+ */
 public class Main {
+    /**
+     * Here starts point of the program
```
This is not correct from the English point of view. The sentence should finish with period. Again, it tells what a programmer already knows about  the `main()` method.

What does the method do? Why are we having it?

alexander-yevsyukov commented on src/main/java/com/example/Main.java:
```diff
@@ -1,7 +1,21 @@
-
+        /**
+         * Object of the Customer class that takes a String
+         */
```
You don't add Javadoc for variables. Instead, you name variables so that it's clear what's inside them.

alexander-yevsyukov reviewed (changes requested):
Please see my comments.

elena-yevs committed:
Merge remote-tracking branch 'origin/class-customer' into class-customer

# Conflicts:
#	src/main/java/com/example/Customer.java

elena-yevs committed:
Changing Javadoc for Customer.java:
1. new description for class Customer.
2. loyaltyCard is final.
3. editing Javadoc for constructor.
Editing Javadoc for Main.java.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,19 @@
+
+/**
+ * A customer is a person who visits a shop for buying goods.
+ * The customer can have a loyaltyCard that has a specific number.
```
The 2nd paragraph must start after an empty line and have `<p>` before the first letter of the text. This is so because it's a paragraph.

alexander-yevsyukov replied:
> `loyaltyCard`

This looks like a name in a problem domain, not a piece of code.

If this is a name in the code it should be either `{@code LoyaltyCard}` or `{@link LoyaltyCard}`. If not, it should go just plain text.

alexander-yevsyukov replied:
Another thing, details of the `LoyaltyCard` type should be described in that class. It could be a number, a QR code, a finger print, or retina scan. Being inside the customer, we don't care. We get an instance of `LoyaltyCard` and we use its API.  Internals must be encapsulated (hidden) as much as possible. This concerns the documentation too.

alexander-yevsyukov replied:
This group of comments was not addressed. Please do.

alexander-yevsyukov replied:
@elena-yevs, the second sentence in Javadoc must start with `<p>`. Please review Javadocs in Guava to understand what it looks like.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,19 @@
+     *
+     * @param loyaltyCard String value
+     */
+    public Customer(String loyaltyCard) {
```
The constructor must accept `LoyaltyCard card`, not a `String`.

alexander-yevsyukov reviewed (changes requested):
Please see my comments.

elena-yevs committed:
Formatting Javadoc for Customer.java.

elena-yevs committed:
Adding LoyaltyCard.java to make changes in parameters of constructor in Customer.java.

elena-yevs committed:
Changes in Customer.java and Main.java:
1. Changing the type of loyaltyCard in Customer.java.
2. The fields of Main.java are deleted.

elena-yevs committed:
Adding gradle-wrapper that contains gradle-wrapper-6.7.1.jar and gradle-wrapper.properties.

alexander-yevsyukov commented on .gradle/buildOutputCleanup/cache.properties:
```diff
@@ -1,2 +1,2 @@
-#Sat Dec 05 22:32:26 EET 2020
-gradle.version=6.5
+#Thu Dec 24 13:00:37 EET 2020
```
As per our Telegram chat, please remove this file. It's not supposed to be in the VCS.

alexander-yevsyukov replied:
The same comment concerns all the files from the `.gradle` directory.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+     * @param id is unique number
+     */
+    public LoyaltyCard(String id);
+
```
Please fix the code layout to standard Java conventions. It's not going to compile, BTW. You've got `;` at the end.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+    /**
+     * Customer's id.
+     */
+    private final String id;
```
Let's call it `number`.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+ */
+public class LoyaltyCard {
+    /**
+     * Customer's id.
```
No, the whole card is the ID of a customer. It's just like passport for a citizen in the context of a country, but simpler, and in the context of a shop.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+
+
+/**
+ * Class implements Value Object.
```
You don't need to refer to a patterns in Javadoc. It's usually assumed by the way a class is implemented. 

You can probably add such things using `@apiNote`, but it's too fine details for the current stage of implementation.

Please describe what the type is about in plain shop-related terms. Try formulating it on paper in Russian, and then translate.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Class implements Value Object.
+ */
+public class LoyaltyCard {
```
Most likely the class is `final`. Assume all the classes you create are `final`, unless you specifically need to plan for inheritance.

alexander-yevsyukov replied:
@elena-yevs, any thoughts on this my comment?

alexander-yevsyukov replied:
@elena-yevs, the comment still applies.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+    private final String id;
+
+    /**
+     * Constructor for checking id with method checkNotNull.
```
You don't need to duplicate what a method or a constructor does in Javadoc. You just document an intent. It should be something like:

```
Creates a new card with the passed number.
```

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+    /**
+     * Constructor for checking id with method checkNotNull.
+     *
+     * @param id is unique number
```
This is the place to document non-nullity. Something like:

```java
  @param number a card number unique for a shop, must be a sequence of ?? number characters
```

Replace `??` with the length of the sequence. BTW, the type passed to the constructor could be `CharSequence` instead of `String`.

alexander-yevsyukov commented on src/main/java/com/example/Main.java:
```diff
@@ -1,7 +1,10 @@
 package com.example;
 
+
 public class Main {
```
I think we can safely remove this class since it became empty.

alexander-yevsyukov reviewed (changes requested):
Please see my comments.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,20 @@
+    private final LoyaltyCard loyaltyCard;
+
+    /**
+     * Constructor for determination the customer's loyalty card number.
```
The word “determination” has a different meaning in English. Try rephrasing to make it read shorter and to the point.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,20 @@
+     * @param loyaltyCard
+     */
+    public Customer(LoyaltyCard loyaltyCard) {
+        this.loyaltyCard = loyaltyCard;
```
You check for non-nullity here too.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,20 @@
+     *
+     * @param loyaltyCard
+     */
+    public Customer(LoyaltyCard loyaltyCard) {
```
There's no need to repeat a type name in a parameter name. Just `card` would be fine.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,20 @@
+    /**
+     * Constructor for determination the customer's loyalty card number.
+     *
+     * @param loyaltyCard
```
If you add Javadoc tag for a parameter you need to document it.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+package com.example;
+
+import static com.google.common.base.Preconditions;
+
```
Please remove empty line here.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,29 @@
+    public String id() {
+        return id;
+    }
+}
```
Please allow for an empty line at the end of the file.

alexander-yevsyukov commented:
> Explanation of the role of the customer and his id (loyaltyCard)

You can say it simpler. Some alternatives to consider:
 * Add `Customer` and `Loyalty` card — here we refer to types.
 * Define customers and loyalty cards — a less formal title that uses plain English terms, but refers to the domain language.

alexander-yevsyukov commented:
In order to move faster, how about dropping the `Customer` type for the time being, and focusing only on `LoyaltyCard`?

elena-yevs committed:
Main.java is deleted.

elena-yevs committed:
Changes in Javadoc;
Rename the variable `id` to `number`;
Adding an empty line;
Code formatting;

elena-yevs committed:
Renaming the field `LoyaltyCard` to `card`;
Javadoc for constructor is deleted.

elena-yevs committed:
`number` that is passed to the constructor is CharSequence.
Changing Javadoc for constructor.

alexander-yevsyukov commented on .idea/.gitignore:
```diff
@@ -1,3 +1,73 @@
+
+# Cache of project
+.gradletasknamecache
+
```
Why do we need so much empty space? Are we reserving space for planting potato or what?

alexander-yevsyukov commented on .idea/.gitignore:
```diff
@@ -1,3 +1,73 @@
+# # suppress inspection "UnusedProperty"
+# org.gradle.java.home=/Library/Java/JavaVirtualMachines/jdk1.8.0_51.jdk/Contents/Home/
+# -------
+gradle.properties
```
Finish files with empty line at the end. It's easy to spot by the red sign at the end. If you review your changes in a PR, it's easy to see.

You can also tune your IDEA to add empty lines at the end of modified files. There is such a setting. Google to your help.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * A customer is a person who visits a shop for buying goods.
+ * <p>
+ * The customer can have a loyalty card that has a specific number.
```
It should be `<p>The customer...`. This is a newer style. Older Java code uses `<p>` as a paragraph delimiters. When we apply modern static code analyzers like Error Prone or Check Style, we'd spot this automatically. For now, please put the paragraph marks before the first character of the text that follows.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,14 @@
+ * The customer can have a loyalty card that has a specific number.
+ */
+public final class Customer {
+    private final LoyaltyCard card;
```
Please put Javadoc here to. Explain why does a customer have a card? 

What happens if a customer does not have a card? Suppose, I visit the shot for the very first time. I surely do not have a card, but I want to purchase a bottle of water because I'm thirsty. Can I?

How does the code behave in such a case?

Also, what happens if I do have a card, but forgot it at my home?

When do we _really_ need the card?

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * Constructor creates a new card with the passed number.
+     *
+     * @param number a card number that is unique, must be a sequence of a certain number of digital characters
```
The current text looks like a word salad. Please try make it _much_ shorter and to the point.

alexander-yevsyukov reviewed (changes requested):
Please see my comments.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,28 @@
+        this.number = number;
+    }
+
+    public String id() {
```
Every `public`, `protected`, and package-private method MUST be documented.

elena-yevs committed:
format .gitignore

elena-yevs committed:
improve Javadoc

elena-yevs committed:
Javadoc for `card`

elena-yevs committed:
improve Javadoc for `number`

elena-yevs committed:
adding Javadoc for `id()`

elena-yevs committed:
adding .gitignore in the root of the project

elena-yevs committed:
editing .gitignore

elena-yevs committed:
ignore the .gradle directory

elena-yevs committed:
ignore all files in the .gradle directory

elena-yevs committed:
setting .gitignore to ignore .bin and .lock files

elena-yevs committed:
ignore more files

elena-yevs committed:
ignoring files and directory

elena-yevs committed:
editing Javadoc for `number`

elena-yevs committed:
improve Javadoc for constructor `Customer()`;
edit Javadoc for class `LoyaltyCard`

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public final class Customer {
+    /**
+     *The customer's loyalty card provides him with bonuses that can be used to pay for purchases.
```
Please have a space character before `The`.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,28 @@
+    /**
+     *The customer's loyalty card provides him with bonuses that can be used to pay for purchases.
+     *
+     * <p>The customer's card number is unchanged.
```
This text belongs to the description of the card. It should not be here. Please remove.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,28 @@
+     *
+     * <p>The customer's card number is unchanged.
+     *
+     * <p>The card number must be entered to receive the bonuses.
```
I'd say it like this: “In order to use previously collected bonuses, the customer uses the card during a purchase”.

This also isolates the “business” of cards and bonuses. It may not be a number, it may be QR code or whatever.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,28 @@
+     * @param card is an additional customer identifier.
+     */
+    public Customer(LoyaltyCard card) {
+        this.card = card;
```
```suggestion
        this.card = checkNotNull(card);
```

alexander-yevsyukov reviewed (changes requested):
Please see my comments. Please go through ALL of them.

elena-yevs committed:
class `LoyaltyCard` is final

elena-yevs committed:
Put a space character before `The` in Javadoc for `Customer`.

elena-yevs committed:
edit `.gitignore` to view files .bin and .lock

elena-yevs committed:
edit `.gitignore` to ignore files .bin and .lock

elena-yevs reviewed (commented):
Files were deleted

elena-yevs committed:
Edit Javadocs.

elena-yevs committed:
Add a method for checking card number

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * A customer is a person who visits a shop for buying goods.
+ *
+ * <p>The customer can have a loyalty card that has a specific number.
```
>  that has a specific number.

These details are local to the `LoyaltyCard` type. It's what encapsulation is all about. You hide the complexity, and do not expose it all over the places in your code :)

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * In order to use previously collected bonuses, the customer uses the card during a purchase.
+     *
+     * <p>The absence of a card does not prevent the customer from purchasing the goods.
```
Please move this documentation into the class header, and have a shorter comment on the card.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @param card is an additional customer identifier.
+     */
+    public Customer(LoyaltyCard card) {
```
The parameter needs to be marked `@Nullable`.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * <p>The card number must be checked.
+     *
+     * @param card is an additional customer identifier.
```
Please remove the dot at the end of the parameter description.

Also, please mention that a customer may not have a card. Refer to my comment on nullability of the parameter.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,36 @@
+
+
+/**
+ * Each customer has his own card that serves as an additional identifier.
```
`his own` — this is a tautology.

`additional` — in addition to what? What kind of identifier does a customer already have?

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public final class LoyaltyCard {
+    /**
+     * Customer's card id.
```
Is it an ID or a number? What's the difference? Why isn't the field named `id` then?

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Constructor creates a new card with the passed number.
+     *
+     * <p>Checking the card number.
```
What did you mean to say? Checking for what?

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,36 @@
+     *
+     * <p>Checking the card number.
+     *
+     * @param number a unique card number, consisting of digital characters.
```
How do we know that the number is unique?

Can we put 100_000 characters into this sequence?

What happens if it's just a couple of characters?

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Method to get the card id if the customer provides a card or its number.
+     *
+     * @return unique card number.
```
Please remove the dot at the end.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    /**
+     * Method to get the card id if the customer provides a card or its number.
```
Of course, it's a method. I see it from the code below. Say it shorter, and to the point.

alexander-yevsyukov commented on src/main/java/com/example/LoyaltyCard.java:
```diff
@@ -0,0 +1,36 @@
+     * @return unique card number.
+     */
+    public String id() {
+        return number;
```
Is it `id()` or a `number()`? Why do we make this gipsy tricks with the reader's mind?

alexander-yevsyukov reviewed (changes requested):
We're progressing. Please see my comments.

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * <p>The customer's card number is unchanged.
+     *
+     * <p>The card number must be checked.
```
By whom? When?

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,30 @@
+    private final LoyaltyCard card;
+
+    /**
+     * Loyalty card is assigned to a specific customer.
```
Specific in what way?

alexander-yevsyukov commented on src/main/java/com/example/Customer.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * <p>The card provides customer with bonuses that can be used to pay for purchases.
+     *
+     * <p>The customer's card number is unchanged.
```
Why do we want these details in the context of this class?

elena-yevs replied:
To describe the purpose of the loyalty card and its feature ( `number is unchanged`).

elena-yevs committed:
remove dots in Javadoc

elena-yevs committed:
remove dots in Javadoc

elena-yevs committed:
edit Javadocs

elena-yevs committed:
edit Javadocs for class `Customer` and field `card`

elena-yevs committed:
`card` is marked as `@Nullable`

elena-yevs committed:
edit Javadoc

elena-yevs committed:
change the name of the method, improve Javadoc

elena-yevs committed:
edit Javadoc for `LoyaltyCard`

elena-yevs committed:
edit Javadoc for `number`

elena-yevs committed:
improve Javadoc for `LoyaltyCard`

alexander-yevsyukov commented on .idea/gradle.xml:
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
+<<<<<<< HEAD
```
This means there is Git merge in progress. Please finish this process.

alexander-yevsyukov reviewed (approved):
LGTM, but please finish the merge.

elena-yevs committed:
delete the conflict markers
